src/tests/cli.cpp

```diff
@@ -601,6 +603,462 @@ TEST_F(rnp_tests, test_cli_rnpkeys)
+                   "--expiration",
+                   "2h",
+                   "--userid",
+                   "expiration_2hours@rnp",
+                   NULL);
+    assert_int_equal(ret, 0);
+
+    ret = call_rnp("rnpkeys",
+                   "--password",
+                   "1234",
+                   "--homedir",
+                   GENKEYS,
+                   "--generate-key",
+                   "--expiration",
+                   "2d",
+                   "--userid",
+                   "expiration_2days@rnp",
+                   NULL);
+    assert_int_equal(ret, 0);
+
+    ret = call_rnp("rnpkeys",
+                   "--password",
+                   "1234",
+                   "--homedir",
+                   GENKEYS,
+                   "--generate-key",
+                   "--expiration",
+                   "2w",
+                   "--userid",
+                   "expiration_2weeks@rnp",
+                   NULL);
+    assert_int_equal(ret, 0);
+
+    ret = call_rnp("rnpkeys",
+                   "--password",
+                   "1234",
+                   "--homedir",
+                   GENKEYS,
+                   "--generate-key",
+                   "--expiration",
+                   "2m",
+                   "--userid",
+                   "expiration_2months@rnp",
+                   NULL);
+    assert_int_equal(ret, 0);
+
+    ret = call_rnp("rnpkeys",
+                   "--password",
+                   "1234",
+                   "--homedir",
+                   GENKEYS,
+                   "--generate-key",
+                   "--expiration",
+                   "2y",
+                   "--userid",
+                   "expiration_2years@rnp",
+                   NULL);
+    assert_int_equal(ret, 0);
+
+    auto         keystore = new rnp_key_store_t(PGP_KEY_STORE_GPG, "");
+    pgp_source_t src = {};
+    assert_rnp_success(init_file_src(&src, GENKEYS "/pubring.gpg"));
+    assert_true(rnp_key_store_load_from_src(keystore, &src, NULL));
+    assert_int_equal(rnp_key_store_get_key_count(keystore), 30);
+    src_close(&src);
+    for (auto &key : keystore->keys) {
+        pgp_key_t *pk;
+        if (key.is_primary()) {
+            pk = &key;
+        } else {
+            assert_true(key.has_primary_fp());
+            pk = rnp_key_store_get_key_by_fpr(keystore, key.primary_fp());
+        }
+        assert_int_equal(pk->uid_count(), 1);
+        auto uid = pk->get_uid(0).str;
+        auto expiration = key.expiration();
+        if (uid == "expiration_absolute@rnp" || uid == "expiration_beyond2038_absolute@rnp" ||
+            uid == "expiration_beyond2038_relative@rnp") {
+            auto requestedExpiration = uid_to_expiration[uid];
+            auto diff = requestedExpiration < expiration ? expiration - requestedExpiration :
+                                                           requestedExpiration - expiration;
+            assert_true(diff < 600); // allow 10 minutes diff
```

Comment:
Asserting in cycle would not quickly tell which exactly key failed. This may be improved, for instance, with some bool function which checks key, and returns false or true depending on checks. Then we'll may assert_true(key_check(keystore, "userid")).